pageserver/src/tenant/timeline.rs

```diff
@@ -3775,6 +3775,8 @@ impl Timeline {
         let mut completed_keyspace = KeySpace::default();
         let mut image_covered_keyspace = KeySpaceRandomAccum::new();
 
+        let mut in_memory_layers_considered = Vec::new();
```

Comment:
I don't think this works. Let's say you have (boxes are in-mem layers, `XXX` is image layer):
```
+-------------------+
|                   |
+-------------------+
+-------------------+
|         XXX       |
+-------------------+
```

You are reading the entire keyspace and get to the second in-mem layer:
* mark it as considered
* set `cont_lsn` to the image layer LSN
* read half the in-mem layer

Now we search for new layers again. The key-range that overlaps the image layer finds it (that's good).
The key ranges that don't overlap the image layer can't see the rest of the in-memory layer and, thus, skips it (that's not good).

This indicates that the in mem layer skipping should be tracked on a per keyspace basis.

Reply:
Thinking about this some more:
All this logic should live in `LayerMap::range_search`:
* Model the in-memory layers as keyspace coverage. This should be fairly easy since these layers cover the whole keyspace.
* Tweak the `RangeSearchCollector` to accept in-mem layer coverage too
* `RangeSearchCollector::add_range` calls into `LayerMap::select_layer` to deal with overlapping layers in the key dimension. It's here that we deal with overlapping deltas and images in the LSN dimension. The logic for in-mem and image layer overlaps should go here too.
* Unit test this till the white smoke blows